pkg/tfgen/generate_csharp.go

```diff
@@ -0,0 +1,1312 @@
+	rg.w.Writefmtln("        public static Task%s %s(%sInvokeOptions? options = null)",
+		typeParameter, methodName, argsParamDef)
+	rg.w.Writefmtln("        {")
+
+	rg.w.Writefmtln("            return Pulumi.Deployment.Instance.InvokeAsync%s(\"%s\", %s, options.WithVersion());",
+		typeParameter, rg.fun.info.Tok, argsParamRef)
+	rg.w.Writefmtln("        }")
+
+	// Close the class.
+	rg.w.Writefmtln("    }")
+}
+
+func (rg *csharpResourceGenerator) generateDatasourceArgs() {
+	if rg.fun.argst != nil {
+		// TODO(pdg): this should pass true for plainType
+		rg.generateInputType(rg.fun.argst, "ResourceArgs", false /*nested*/, false /*plainType*/)
+	}
+}
+
+func (rg *csharpResourceGenerator) generateDatasourceResult() {
+	if rg.fun.retst != nil {
+		rg.generateOutputType(rg.fun.retst, false /*nested*/)
+	}
+}
+
+func (rg *csharpResourceGenerator) generateInputProperty(prop *variable, nested, plainType bool) {
+	qualifier := ""
+	if !nested {
+		qualifier = "Inputs"
+	}
+
+	wireName := prop.name
+	propertyName := csharpPropertyName(prop)
+	propertyType := qualifiedCSharpPropertyType(prop, qualifier, !plainType /*wrapInput*/)
+
+	// First generate the input attribute.
+	attributeArgs := ""
+	if !prop.optional() {
+		attributeArgs = ", required: true"
+	}
+	if rg.res != nil && rg.res.IsProvider() && prop.typ.kind != kindString {
+		attributeArgs += ", json: true"
+	}
+
+	// Next generate the input property itself. The way this is generated depends on the type of the property:
+	// complex types like lists and maps need a backing field.
+	switch prop.typ.kind {
+	case kindList, kindSet, kindMap:
+		backingFieldName := "_" + prop.name
+		backingFieldType := qualifiedCSharpType(prop.typ, qualifier, !plainType /*wrapInput*/)
+
+		rg.w.Writefmtln("        [Input(\"%s\"%s)]", wireName, attributeArgs)
+		rg.w.Writefmtln("        private %s? %s;", backingFieldType, backingFieldName)
+
+		if hasDocComment(prop) {
+			rg.w.Writefmtln("")
+		}
+		emitCSharpPropertyDocComment(rg.w, prop, "        ")
+
+		rg.w.Writefmtln("        public %s %s", propertyType, propertyName)
+		rg.w.Writefmtln("        {")
+		rg.w.Writefmtln("            get => %s ?? (%s = new %s());", backingFieldName, backingFieldName, backingFieldType)
+		rg.w.Writefmtln("            set => %s = value;", backingFieldName)
+		rg.w.Writefmtln("        }")
+	default:
+		initializer := ""
+		if !prop.optional() {
+			initializer = " = null!;"
+		}
+
+		emitCSharpPropertyDocComment(rg.w, prop, "        ")
+		rg.w.Writefmtln("        [Input(\"%s\"%s)]", wireName, attributeArgs)
+		rg.w.Writefmtln("        public %s %s { get; set; }%s", propertyType, propertyName, initializer)
+	}
+}
+
+func (rg *csharpResourceGenerator) generateInputType(typ *propertyType, baseType string, nested, plainType bool) {
+	contract.Assert(typ.kind == kindObject)
+
+	rg.w.Writefmtln("")
+
+	// Open the class.
+	if baseType != "" {
+		baseType = " : " + baseType
+	}
+	rg.w.Writefmtln("    public class %s%s", typ.name, baseType)
```

Comment:
might be good to seal these.

Reply:
this also applies to our other classes we generate.

Reply:
Although... let's not seal our resources just yet.  we have realistic cases where we use subclassing to make it easier to create a resource.